spring-cloud-commons/src/test/java/reactor/core/publisher/FluxFirstNonEmptyEmittingTests.java

```diff
@@ -0,0 +1,141 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package reactor.core.publisher;
+
+import java.util.Arrays;
+
+import org.junit.Test;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscription;
+import reactor.core.CoreSubscriber;
+import reactor.core.Scannable;
+import reactor.test.StepVerifier;
+
+import static java.util.Collections.singletonList;
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * @author Tim Ysewyn
+ */
+public class FluxFirstNonEmptyEmittingTests {
+
+	@Test(expected = NullPointerException.class)
+	public void arrayNull() {
+		CloudFlux.firstNonEmpty((Publisher<Integer>[]) null);
+	}
+
+	@Test(expected = NullPointerException.class)
+	public void iterableNull() {
+		new FluxFirstEmitting<>((Iterable<Publisher<Integer>>) null);
+	}
+
+	@Test
+	public void firstWinner() {
+		StepVerifier
+				.create(CloudFlux.firstNonEmpty(Flux.range(1, 10), Flux.range(11, 10)))
+				.expectNext(1, 2, 3, 4, 5, 6, 7, 8, 9, 10).verifyComplete();
+	}
+
+	@Test
+	public void firstWinnerBackpressured() {
+		StepVerifier
+				.create(CloudFlux.firstNonEmpty(Flux.range(1, 10), Flux.range(11, 10)))
+				.thenRequest(5).expectNext(1, 2, 3, 4, 5).thenCancel()
+				.verifyThenAssertThat().hasNotDiscardedElements().hasNotDroppedElements()
+				.hasNotDroppedErrors();
+	}
+
+	@Test
+	public void secondWinner() {
+		StepVerifier
+				.create(CloudFlux.firstNonEmpty(Flux.never(), Flux.range(11, 10).log()))
+				.expectNext(11, 12, 13, 14, 15, 16, 17, 18, 19, 20).verifyComplete();
+	}
+
+	@Test
+	public void secondEmitsError() {
+		RuntimeException ex = new RuntimeException("forced failure");
+		StepVerifier
+				.create(CloudFlux.firstNonEmpty(Flux.never(), Flux.<Integer>error(ex)))
+				.thenCancel().verifyThenAssertThat().hasNotDiscardedElements()
+				.hasNotDroppedElements().hasNotDroppedErrors();
+	}
+
+	@Test
+	public void singleArrayNullSource() {
+		StepVerifier.create(CloudFlux.firstNonEmpty((Publisher<Object>) null))
+				.expectError(NullPointerException.class).verify();
+	}
+
+	@Test
+	public void arrayOneIsNullSource() {
+		StepVerifier.create(CloudFlux.firstNonEmpty(Flux.never(), null, Flux.never()))
+				.expectError(NullPointerException.class).verify();
+	}
+
+	@Test
+	public void singleIterableNullSource() {
+		StepVerifier
+				.create(CloudFlux.firstNonEmpty(singletonList((Publisher<Object>) null)))
+				.expectError(NullPointerException.class).verify();
+	}
+
+	@Test
+	public void iterableOneIsNullSource() {
+		StepVerifier
+				.create(CloudFlux.firstNonEmpty(Arrays.asList(Flux.never(),
+						(Publisher<Object>) null, Flux.never())))
+				.expectError(NullPointerException.class).verify();
+	}
+
```

Comment:
this notably lacks any test that reflects what distinguish this operator from `Flux#first`: multiple tests that validate an early completing source like `Flux.empty()` vs other source(s) that emit an element. Testing the case where none of the sources emit any element would be interesting as well.